[vcpkg] RFC: Manifests

strega-nil opened the pull request:
We are currently working on a new feature: manifests. They are similar to manifests from other languages: Cargo.toml, package.json, etc. This specification lays out how we've thought about designing the manifest so far; please add your comments so that we might have the best experience possible :)

A note: there are two major features that pretty much depend on manifests, but which will be defined in separate RFCs: versioning, and federation. This manifest is designed such that those features are possible to fit on top of it, but it does not attempt to define those features right now.

Currently, this feature is being developed at [strega-nil/vcpkg/manifests]; please come check it out for very very pre-alpha demo-friendly support of this! An example of this feature in use is at [strega-nil/vcpkg-example] 😄 

[rendered]

[strega-nil/vcpkg/manifests]: /strega-nil/vcpkg/tree/manifest
[strega-nil/vcpkg-example]: /strega-nil/vcpkg-example
[rendered]: /strega-nil/vcpkg/blob/manifest/docs/specifications/manifests.md

strega-nil committed:
[vcpkg] Add manifests specification

Edhebi commented:
It looks cool! I don't have any constructive feedback regarding the manifest itself, but the nested markdown lists are broken in the `Definitions` section.

I can submit a fix if you want. I suppose I should PR your own branch for that ?

strega-nil commented:
yeah, that works! sorry I can't work on it right now, I'm doing work on macOS pre-10.15 support -.-

petersteneteg commented:
Cool! so looking forward to this

One question the `vcpkg_modules` folder seems to go in <manifest-root> which I assume would usually be the root directory of the project itself. To make sure the manifest will be version controlled. But when using CMake the best practices now days seems to be to keep the project folder clean of any build product, ie. out of source builds. For the example in presented one would usually have:
```
projdir/
  build/
  example/
    src/
      main.cxx
    CMakeLists.txt
    vcpkg.json
```
or something like that. I would prefer if the `vcpkg_modules` folder was either in the build folder or at the same level as the build folder. At least I would never want it to be inside of example.

traversaro commented on docs/specifications/manifests.md:
```diff
@@ -0,0 +1,312 @@
+mode.
+
+When in "modern" mode, the `installed` directory will be changed to
+`<manifest-root>/vcpkg_modules` (name up for bikeshedding).
```
The "legacy" mode has a name that seems to suggest that it will eventually be removed or actively discouraged. However, the use case for the "modern" mode is providing the dependency for a single project. This seems to be quite different  from at least one widespread  use case of "legacy" mode, i.e. providing a set of shared dependency for several projects that then are build together, for example using a [`build_tool`](https://design.ros2.org/articles/build_tool.html) such as [colcon](https://colcon.readthedocs.io/en/released/) or [CMake-based superbuilds](https://blog.kitware.com/cmake-superbuilds-git-submodules/).

Just to understand, is the use case of providing dependency for different projects being explicitly considered "legacy" in vcpkg?

strega-nil replied:
As of now we don't plan to deprecate or stop supporting the existing style of using vcpkg; this is just a new way of using vcpkg, one that we think is a better experience for a lot of our customers, but it is not the only one. Just one of the two primary usecases :)

I personally like "legacy" vs "modern", since, at least to me, "legacy" doesn't have a negative connotation; however, if you have another suggestion..?

traversaro replied:
Full disclosure: I am not a native English speaker, so the meaning that I associate of "legacy" may be wrong, a quick search on my Google as first answer gives (see https://www.lexico.com/en/definition/legacy): 
> Denoting or relating to software or hardware that has been superseded but is difficult to replace because of its wide use.

However, if it is documented that the "legacy" mode is not "superseded", then I am perfectly ok. 

Just for the love of bikeshedding, a possible alternative to "legacy" is "classic". Yes another alternative is to use terms that do not refer to when the feature was added (then may be outdated if both modes will be used also in the future), and  refer to the feature of the mode itself. For example, we could have the "standalone" and the "manifest" mode, or something similar.

strega-nil replied:
Hmm! I think I like "classic" mode better too :)

strega-nil commented:
@petersteneteg I simply followed existing practice for Javascript; however, there's a good point there that, for compiled languages, it doesn't make sense to commit your `vcpkg_modules` directory. However, it's unclear where any specific build system (or even user! it's user-chosen with build systems like cmake) will place their build trees, and thus `vcpkg install` wouldn't be able to put its `vcpkg_modules` folder underneath the build tree. Thus, I'm not sure of how to do this otherwise...

traversaro commented on docs/specifications/manifests.md:
```diff
@@ -0,0 +1,312 @@
+
+add_executable(example src/main.cxx)
+
+vcpkg_acquire_dependencies()
```
While this may be convenient for some developers, it will break compilation for anyone that wants to build the library without vcpkg (or even with vcpkg, but without  using the vcpkg's toolchain) and providing its own version of the dependencies, making life complicated for packages of systems different from vcpkg (see https://youtu.be/sBP17HQAQjk?t=861 for more on this). 

To avoid  this, could it make this conditional on the toolchain being used, i.e.: 
~~~cmake
if(VCPKG_TOOLCHAIN)
   vcpkg_acquire_dependencies() 
endif()
~~~

or even better, if this block is supposed to be always the same, could it make sense to just move it in the toolchain itself, for example guarded under an option?

strega-nil replied:
If a project wishes to support non-vcpkg dependencies, my expectation is that they will be able to set that up via their own CMake build system; I don't think this should be the job of the function.

traversaro replied:
This make sense. However,  now that I think a bit more on this, my major concern is  not on the function, but rather on the example. If the specification will become part of the public documentation of vcpkg, it will appear that this is the recommended way of using vcpkg in a CMake project. To avoid that new inexperienced users build projects that hardcode vcpkg use, it would be great if there was a mention in the docs that if you want to create a "packagable library" hardcoding a single package  manager  in the build system may not be the best option.

traversaro commented on docs/specifications/manifests.md:
````diff
@@ -0,0 +1,312 @@
+```
+
+This function first acquires, then `find_package`s, the
+dependencies.
````
This will work by assuming that by maintaining some kind of mapping between the  vcpkg's port name and the corresponding CMake package name somewhere?

strega-nil replied:
Yes; I figure it'll become obvious in the implementation how this should work, and I'll backfill details later. This specific part is not a fully fleshed out design 😄

Neumann-A replied:
The mapping could be included in the `vcpkg.json` as `cmake_package_name` or similar. But I am against vcpkg defining custom targets since those are not `portable`

strega-nil replied:
I don't want us to define custom targets; it would only be for libraries which give you an easy target to link against.

edhebi committed:
Remove invisible tabs and fix nested lists

Somehow, a lot of tabs where inserted betweend the `*` and start of list items, and they didn't show up in the github source view 🤷.

traversaro commented on docs/specifications/manifests.md:
```diff
@@ -0,0 +1,312 @@
+
+## Specification
+
+A manifest file shall have the name `vcpkg.json`, and shall be in the root directory of a package.
```
What does `package` means in this context? A CMake package? Or a git repository? In some context (for example ROS packages in the robotics community) it is quite common to have a  single repository that contains multiple "CMake packages", see for example https://github.com/ros2/rclcpp .

strega-nil replied:
It's unclear how to define package, and this is why I don't try to. It's basically "the place you put a `vcpkg.json` file"

traversaro reviewed (commented):
Thanks a lot for working on this! I have added  a few comments inline, I hope they can be useful.

traversaro commented:
> Additionally, I don't think that out-of-source-tree builds are generally accepted practice with modern CMake. 

If out-of-source-tree builds means "keeping the build directory not inside the source directory", I don't think it is uncommon.  Probably it is just my bubble, but out-of-source-tree builds are standard practice when using [build_tools](https://design.ros2.org/articles/build_tool.html) such as [colcon](https://colcon.readthedocs.io/en/released/) or [CMake-based superbuilds](https://blog.kitware.com/cmake-superbuilds-git-submodules/). Furthermore, it is what you get by default when consuming dependencies via [CMake's FetchContent](https://cmake.org/cmake/help/v3.11/module/FetchContent.html),  and if i am not wrong  vcpkg itself builds its own ports out-of-source-tree when using `vcpkg_configure_cmake`.

strega-nil commented:
@traversaro sorry, I meant the other way around -- it's not that out-of-source-tree builds aren't accepted practice, but that in-source-tree builds are totally valid.

kobalicek commented:
Feedback:
  - You need revisions - since package information can be maintained out of project there has to be some information about the revision. Imagine you fix vcpkg, but not the main project, the revision of the package should change, but it's still for the same version of the library.. Conan has this, for example.
  - Out of source build should be the only build possible. I'm not even sure CMake still allows in-source builds - nobody really wants in-source builds these days anyway, it's a bad practice.
  - Each directory one vcpkg? This is backwards - a nightmare everybody faced when having large cmake projects, for example, where you have 100 CMakeLists.txt files - find the right one in an editor, find the right one in 20 open tabs...
  - I think it would be nice if vcpkg was unobtrusive - I don't like it couples with CMakeLists - it gives users no choice, which is something that I won't accept, but I may still provide vcpkg.json in my project for people that want to use it. So it would be cool if it worker like other package managers - you install dependencies, then you build do project, but you won't do both at the same time.

strega-nil commented:
@kobalicek 

* `port-version` is the way to do this
* I disagree strongly; it is very much not bad practice, and it's literally the recommended way to build CMake. I have no idea where people got this idea from.
* You don't... have to have multiple `vcpkg.json`s in a project; in fact this initial specification does not support multiple `vcpkg.json`s in a project.
* It doesn't have to couple with CMakeLists.txt, that's just one possible usecase. I should add an example of a project that doesn't couple with CMakeLists.txt, I guess.

strega-nil committed:
feedback from @traversaro

petersteneteg commented:
> Edit: okay, someone explained to me what "in-source builds" are. Sorry for being confusing! I'm not trying to argue that in-source builds are good, but that in-project builds are good. I hadn't even considered in-source builds to be a thing that one might do... the fun of being young, I guess!
But why put the vcpkg_modules folder "in-source" then? the vcpkg_acquire_dependencies() should know about the build directory...

strega-nil commented:
@petersteneteg The `vcpkg_modules` is not in-source, just in-project. Quite different :)

petersteneteg commented:
> @petersteneteg The `vcpkg_modules` is not in-source, just in-project. Quite different :)

Citing Professional CMake: A Practical Guide https://crascit.com/professional-cmake/ 
Chapter 2.2

> This book will always use out-of-source builds and will follow the pattern of the source and build directories being under a common parent. The build directory will be called build, or some variation thereof. 

> A variation on this used by some developers is to make the build directory a subdirectory of the source directory. This offers most of the advantages of an out-of-source build, but it does still carry with it some of the disadvantages of an in-source arrangement. Unless there is a good reason to structure things that way, keeping the build directory completely outside of the source tree instead is recommended.

Hence I would strongly prefer to keep the vcpkg_modules folder completely outside of the source directory. And by source directory I mean the root of my repository and the same directory as were the manifest file would be in.

kobalicek commented:
I think they should look into Conan - not saying Conan is the best tool, but it solves a lot of problems and maintains its own cache that is configurable through `$CONAN_USER_HOME` - so by having different env-vars you can pretty much isolate it. It has much more additional things that are very usable - like propagating environment variables, library directories, binaries, etc... through projects.

strega-nil commented:
@petersteneteg That's not what that quote is saying, I don't think: the "source directory" is not the directory which the project is in. The source directory is the directory containing the literal source files -- `.c`, `.cxx`, etc. -- usually called `src` or `source`. The build directory and the source directory are distinct, and are in the same directory as any manifests or main build files (`CMakeLists.txt`, for example). I can't speak to that specific book, as I don't really want to buy it and I don't personally have access to it, but I think it's important to realize that, for example:

* vcpkg recommends builds at the same level as `CMakeLists.txt`
* ninja-build recommends builds at the same level as `CMakeLists.txt`
* CMake recommends builds at the same level as `CMakeLists.txt`
* LLVM recommends builds at the same level as `CMakeLists.txt`

so it's extremely common practice to have a project tree that looks like:

```
src/
  <source>...
build/
  <build>...
README.md
CMakeLists.txt
```

petersteneteg commented:
CMake defines the "Source tree" as the directory with the top level CMakeLists.txt. No other parent directory would really matter for CMake. https://cmake.org/cmake/help/v3.17/manual/cmake.1.html#manual:cmake(1) 
For the "Build Tree" is says:
> To maintain a pristine source tree, perform an out-of-source build by using a separate dedicated build tree. An in-source build in which the build tree is placed in the same directory as the source tree is also supported, but discouraged.

Here I interpret _separate_ as not a subfolder. 

The Book does the same interpretation. 
![Skärmavbild 2020-05-08 kl  09 42 37](https://user-images.githubusercontent.com/3638222/81383437-56346500-9110-11ea-96dd-d8d4147e7f3e.png)
 
Using a subfolder might be common practice. But that does not make it good practice. 
And it would be unfortunate if vcpkg enforced that pattern.

kobalicek commented:
I don't really see a difference between:

```
src/
  <source>...
build/
  <build>...
README.md
CMakeLists.txt
```

and

```
src/
  CMakeLists.txt
  <source>...
build/
  <build>...
README.md
```

Both are basically out-of-source builds, because the build tree doesn't intermix with source-tree. Every project will decide where the CMakeLists.txt will be, so `vcpkg` shouldn't care anyway.

meastp commented:
Thank you for working on this!

Am I correct in assuming that this could also be used even if one only uses `vcpkg` to manage/acquire third-party dependencies? 

For example, we develop in Visual Studio and use MSBuild to build our project(s)...

Our use case:
```
myprojects/ //project root with multiple projects
  proj1/
  proj2/
  proj3/
  x64/Release/ // custom build output folder for all .exe .dll and third-party binaries
  myprojects.sln
  vpkg.json // all dependencies for all projects
```

I'm guessing that this works with `vcpkg integrate install`, but how are different versions of packages handled and how will this work with binarycaching #11204?

Ideally I'd like to create a `vcpkg.json` file that is used during CI and on developer machines to acquire third-party dependencies and use binarycaching to ensure that they all use the same version and avoid building from scratch every time.

strega-nil commented:
@meastp yep, absolutely! before this is stabilized, we will support every build system we support currently 😁

petersteneteg commented on docs/specifications/manifests.md:
````diff
@@ -0,0 +1,349 @@
+```
+
+and we're done! `fmt` will get installed into
+`example/vcpkg_modules`, and we can run our executable with:
````
Does this mean that vcpkg will build and cache the dependency in $VCPKG_ROOT and only install it into vcpkg_modules? So the build can be shared? or would different project will the same $VCPKG_ROOT have to rebuild?

strega-nil replied:
Yeah, vcpkg will cache dependencies in `$VCPKG_ROOT`.

petersteneteg replied:
Awesome!

meastp commented:
> 
> 
> @meastp yep, absolutely! before this is stabilized, we will support every build system we support currently 😁

That's great!

How will this work with custom ports (for example binary third-party library ports I created in a local/remote fork of vcpkg that aren't upstream)?

david-antiteum commented:
It seems that vcpkg.json will have a dual role, when used on a port and when used by a developer to list dependencies of an application. In the second case, the name and version are not required or available elsewhere (for example, in the project entry CMakeLists.txt).

strega-nil commented:
@david-antiteum it's important to note that it's not actually a dual role -- it's just expanding what a "port" looks like.

meastp commented:
@strega-nil We have custom ports that are not present in the official vcpkg repository. Could one point to where (github url, file share etc.) the port "definition" is?

Perhaps this is a separate PR/item?

For example, we have both open source ports and binary proprietary ports... :)

strega-nil commented:
@meastp for this initial RFC, we aren't going to add that feature; however, it's planned for very soon afterwards :)

meastp commented:
@strega-nil Ah, ok, that's fair :)

I currently use a fork of vcpkg where I have added the ports missing from official vcpkg (binary and open source) that we need. What I think you're saying is that:

- before the package federation feature is done, I'll need to keep using my fork and initiating vcpkg.exe (to build dependencies) from my fork (that I have downloaded/cloned and built locally) when working with vcpkg.json in `myprojects` (example above)

- when the package federation feature is done, I won't be dependent on managing my fork at all as long as I can point to the correct path for each dependency (url, path etc) in vcpkg.json in `myprojects`.

this ties in nicely with vcpkg.exe being included in Visual Studio - then I can use that even though I use dependencies that are not part of the official vcpkg repo.

_Are my assumptions correct?_ :)

strega-nil commented:
Your assumptions are correct! Although, if you want to avoid maintaining a fork, you may want to look into `--overlay-ports` until such point as we add federation to the manifest.

daniel-piro commented:
Thanks for working on this!

If I understand correctly, the vcpkg_modules folder will be created for each cmake project and contain the binaries of it’s versioned dependencies. Will there be a way to first check whether a version is already installed “globally” by VCPKG to avoid having the same copy of the library on each dependent project?

strega-nil commented:
@daniel-piro that's not a feature I'm interested in adding personally; I think that having the installed directory be completely different will allow one to separate the two really cleanly. For what it's worth, we will support binary caching, which means if you already have it installed _anywhere_, you'll be able to cache binaries, so that means it's just a zip open -- not a compile.

strega-nil committed:
fix stuff from review

strega-nil commented:
I'm going to Final Comment Period this! This manifest will be merged on Monday, June 1 (a week) if no one else has any comments :)

Please add any other comments you wish to! Thanks all!

nightlark commented:
This is great! It looks like manifests and federation will help alleviate the biggest issue we've run into, which is challenges submitting a PR to the vcpkg repository for ports that we want other people to use.

seanyen commented:
@strega-nil Just curious, for the `vcpkg.json` manifest, does it support (or plan to support) the following workflow?

1. create an empty folder as the root for a new project.
2. create a `vcpkg.json` under it.
3. do `vcpkg install` under the folder and magically `vcpkg` will build and produce the `installed` for this new project?

I am viewing this manifest ting from my experience in `npm`, `pip` and `conda`, and hope the `vcpkg` workflow won't be deviated too much from those popular package managers too. And btw, this is a great work and happy to see `vcpkg` is moving toward this direction.

strega-nil commented:
@seanyen this is absolutely a valid thing to do.
